logbook-ktor-server/src/main/kotlin/org/zalando/logbook/server/LogbookServer.kt

```diff
@@ -32,25 +29,37 @@ class LogbookServer(
     companion object : BaseApplicationPlugin<Application, Config, LogbookServer> {
         private val responseProcessingStageKey: AttributeKey<Logbook.ResponseProcessingStage> =
             AttributeKey("Logbook.ResponseProcessingStage")
+        private val proceedWithKey: AttributeKey<ByteReadChannel> =
+            AttributeKey("Logbook.proceedWith")
+
         override val key: AttributeKey<LogbookServer> = AttributeKey("LogbookServer")
+
         override fun install(pipeline: Application, configure: Config.() -> Unit): LogbookServer {
             val config = Config().apply(configure)
             val plugin = LogbookServer(config.logbook)
 
-            pipeline.receivePipeline.intercept(ApplicationReceivePipeline.Before) {
+            // This interceptor is executed before ApplicationReceivePipeline,
+            // and regardless of whether the request has body or not
+            pipeline.intercept(ApplicationCallPipeline.Call) {
                 val request = ServerRequest(call.request)
                 val requestWritingStage = plugin.logbook.process(request)
-                val proceedWith = when {
-                    request.shouldBuffer() && !call.request.receiveChannel().isClosedForRead -> {
-                        val content = call.request.receiveChannel().readBytes()
-                        request.buffer(content)
-                        ByteReadChannel(content)
-                    }
 
-                    else -> it
+                val proceedWith = if (!call.request.receiveChannel().isClosedForRead) {
+                    val content = call.request.receiveChannel().readBytes()
+                    request.buffer(content)
+                    ByteReadChannel(content)
+                } else {
+                    ByteReadChannel.Empty
                 }
+
                 val responseProcessingStage = requestWritingStage.write()
                 call.attributes.put(responseProcessingStageKey, responseProcessingStage)
+                call.attributes.put(proceedWithKey, proceedWith)
```

Comment:
Do we actually need this attribute ? Isn't it still enough to add the content in the request buffer ?
Changing the type of interceptor might be enough for `responseProcessingStageKey` to always be filled inside the other one :)

Reply:
I really struggled and couldn't due to the way Ktor is design. Please read the PR description carefully as why I did it the way I did.

Needless to say, I may have missed something, so if you have a concrete suggestion, by all means go ahead and add a comment.

Reply:
Well I agree Ktor design seems a little weird when it comes to creating plugins.. They simplified it a little bit in the new API but reading request/response bodies is still a headache.
I will experiment using the [`ResponseObserver`](https://api.ktor.io/ktor-client/ktor-client-core/io.ktor.client.plugins.observer/-response-observer.html) plugin to deal with the response (which could help avoid using internal API from `request.content`.

Back to your question, I'm just wondering if you have spimply tried to test it without having this attribute and without the `ApplicationReceivePipeline.Transform` phase interceptor ?

Now that you managed to have the `responseProcessingStageKey` always added wether or not a body is received, this might not be useful anymore.

Reply:
Hello @msdousti I have pushed a sample here: ([url](https://github.com/grassehh/issues/blob/main/zalando-logbook-1627/src/test/kotlin/com/grassehh/ServerTest.kt))

Basically, I only changed the intecepted phase to `ApplicationCallPipeline.Monitoring` and removed the `proceedWith` which is not necessary and it seems to work.

Reply:
@grassehh 
I replaced my code with your `LogbookServer`, and ran the tests. Two tests fail as follows:

![image](https://github.com/zalando/logbook/assets/3616518/01b0e054-7436-45ea-8952-3989244b28c4)

Reply:
Basically, if you don't use `proceedWith`, the request body is consumed by th "request logger", and it is no longer visible down the pipeline.

Reply:
I see @msdousti you are right. So I found a way by installing the [DoubleReceive](https://ktor.io/docs/server-double-receive.html) inside the custom plugin, and then simply calling `call.receiveChannel().readBytes()` inside the interceptor in order to read the request body.
I updated my sample and added the `LogbookServerTest` and they all pass now :)

Reply:
Thanks, good to know about this plugin!

I've a question though: Is it worth adding another dependency to the project, and add another plugin to the pipeline, while my code is essentially doing that without all the fuss?

This is not a rhetorical question, I'm genuinely interested in the answer. The `DoubleReceive` plugin makes the code somewhat more readable, but the extra dependency is a downside, and the extra plugin **may** also make it less performant/more memory hungry (not sure).

Reply:
I understand. In my opinion I can see several reason of using that plugin:
- It is an official one, thus it could might make it easier to maintain the custom plugin
- Ktor is a complex and sensitive machinery so we don't know if there could be any side-effect or things we didn't think of about reading requests bodies. As an example you can check out my other [issue](https://github.com/zalando/logbook/issues/1822) in the `logbook-ktor-client` that can lead to some deadlocks when reading the response.

Now if you really don't want to add this dependency, which I can understand, it is worth taking a look at the `DoubleReceive` implementation.
Also, could it be a requirement for the users to add the plugin themselves ? We could eventually have two versions of the `logbook-ktor-server`. One that is fully custom and another that relies on the `Ktor` official plugin.

Reply:
Thanks for the explanations. I trust if we load the plugin and the user loads it independently, there will be no conflicts? 

@kasmarian Could you chime in here: Comparing two approaches above (with or without additional plugin), which one do you prefer?

Reply:
I agree with @grassehh on this. The `DoubleReceive﻿` plugin seems like a good way to go. Overall Logbook has to overcome issues related to double-reads of request/response streams in various modules, and here the native plugin provides a solution out of the box.

Reply:
Thanks! Then I'll make the necessary changes and push sometime tomorrow.